File: solutions/cpp/1870/solution.cpp
class Solution {
public:
    double limit;
    int minSpeedOnTime(vector<int>& dist, double hour) {
        if (dist.size() > ceil(hour)) {
            return -1;
        }

        this->limit = hour;
        int left = 1;
        int right = pow(10, 7);
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (check(mid, dist)) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }

    bool check(double speed, vector<int> &dist) {
        double h = 0;
        for (double d : dist) {
            h = ceil(h);
            h += d / speed;
        }

        return h <= limit;
    }
};

File: solutions/golang/1870/solution.go
var limit float64
func minSpeedOnTime(dist []int, hour float64) int {
    if len(dist) > int(math.Ceil(hour)) {
        return -1
    }

    limit = hour
    left := 1
    right := int(math.Pow(10, 7))
    for left <= right {
        mid := left + (right - left) / 2
        if check(float64(mid), dist) {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }

    return left
}

func check(speed float64, dist []int) bool {
    var actualHours float64
    actualHours = 0
    for _, dt := range dist {
        fdt := float64(dt)
        actualHours = math.Ceil(actualHours)
        actualHours += fdt / speed
    }
    return actualHours <= limit
}
